power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```
@@ -714,52 +766,174 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
+            return result;
+        }
+
+        auto const all_comp_update_independence = check_components_independence(update_data);
+        for (auto const& comp : all_comp_update_independence) {
+            result.emplace_back(comp.name, comp.is_independent());
+        }
+
+        return result;
+    }
+    void validate_update_data_independence(UpdateCompProperties const& comp) const {
+        if (!comp.has_id && comp.ids_all_na) {
+            return; // empty dataset is still supported
+        }
+        if (comp.elements_in_base < comp.elements_ps_in_update) {
+            throw DatasetError("Update data has more elements per scenario than input data for component " + comp.name +
+                               "!");
+        }
+        if (comp.ids_part_na) {
+            throw DatasetError("Part of the IDs are not valid for component " + comp.name + " in update data!");
+        }
+        if (!comp.uniform) {
+            if (comp.is_columnar && !comp.has_id) {
+                throw DatasetError("Columnar input data without IDs for component " + comp.name + " is not uniform!");
+            }
+            if (!comp.is_columnar && comp.ids_all_na) {
+                throw DatasetError("Row based input data with all NA IDs for component " + comp.name +
+                                   " is not uniform!");
+            }
+        }
+        if (comp.elements_in_base != comp.elements_ps_in_update) {
+            if (comp.is_columnar && !comp.has_id) {
+                throw DatasetError("Columnar input data for component " + comp.name +
+                                   " has different number of elements per scenario in update and input data!");
+            }
+            if (!comp.is_columnar && comp.uniform && (comp.has_id && comp.ids_all_na)) {
+                throw DatasetError("Row based input data for component " + comp.name +
+                                   " has different number of elements per scenario in update and input data!");
```

i think you can throw the same error twice here.

---

Agreed to resolve.